docs/decisions/backend/0009-advanced-installer-design.md

```diff
@@ -0,0 +1,246 @@
+
+```text
+Red Hair
+├─ Data
+│  ├─ model.dae
+│  └─ hair.dds
+Green Hair
+├─ Data
+│  ├─ model.dae
+│  └─ hair.dds
+Blue Hair
+└─ Data
+   ├─ model.dae
+   └─ hair.dds
+```
+
+The user is expected to select items from one of the given folders. These folders do map to game folder, but
+because there are duplicates and multiple candidates for deployment, the Nexus App cannot automatically
+choose which one to deploy.
+
+## Considered Options (Deployment)
+
+- Reusing the FOMOD Deployment System.
+- Programming a new Deployment System from scratch.
+
+## Decision Outcome
+
+After some investigation work, I it has been concluded that it isn't really possible to reuse the FOMOD system here.
+
+While the FOMOD system is technically a subset of the Advanced Installer if you think about it (after all, toggling
+1 FOMOD option is just like mapping 1-* files to a given location); the design of the current FOMOD code does not
+easily lend itself to reuse.
+
+This is because FOMOD library, emits a list of files to be deployed from the library as 'Instructions'; and we simply
+use those instructions to modify an enumerable of files to be deployed. This file list is created by user interacting with
+the FOMOD library in an asynchronous call (i.e. logic abstracted away by library). As a result, reusing existing FOMOD
+logic does not translate well to `Advanced Installer`,
+
+On the other hand, a good chunk of the UI styling/logic should be reusable.
+In other words, we're going for a custom solution.
+
+### Consequences
+
+- Good, because custom solution allows for easier serialization of the data.
+- Good for maintenance, as the logic for persisting deployment steps could be reused in the FOMOD code.
+    - Note: Both are a file->file mapping effectively, so we can use serialization format for Advanced Installer in FOMOD.
+
+### Implementation Details / Notes
+
+#### Background Knowledge
+
+Currently Mod Installers work by returning an `IEnumerable<AModFile>`, which is a flat collection of files
+to be moved from one location to another.
+
+To use FOMOD as an example, extracting a file from an individual archive works in the following manner:
+```csharp
+return new FromArchive
+{
+    Id = ModFileId.New(),
+    To = new GamePath(gameTargetPath.Type,
+        gameTargetPath.Path.Join(RelativePath.FromUnsanitizedInput(instruction.destination))),
+    Hash = file.Value.Hash,
+    Size = file.Value.Size
+};
+```
+
+#### Emitting Operations
+
+There may be temptation to dynamically emit the extract (`FromArchive`) instructions as the user selects the install
+target in the UI:
+
+![](./images/0009-advanced-installer-location.png)
+
+This however, is not a good idea; as `undo` operations and unmaps are more difficult to process in a solution like this.
+(Might also be costly in huge mods).
+
+Instead, we should maintain a mapping of `Source` -> `Target`, as a dictionary; and only emit the needed operations 'Just in Time'
+at the deployment stage.
```

Comment:
I need some clarification about what you mean with deployment stage

Reply:
Flattening the dictionary into a list of files to deploy in `IModInstaller`